pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/IngestionDelayTracker.java

```diff
@@ -83,22 +90,38 @@
  *
  * TODO: handle bug situations like the one where a partition is not allocated to a given server due to a bug.
  */
-
 public class IngestionDelayTracker {
 
-  private static class IngestionInfo {
-    final long _ingestionTimeMs;
-    final long _firstStreamIngestionTimeMs;
-    final StreamPartitionMsgOffset _currentOffset;
-    final StreamPartitionMsgOffset _latestOffset;
+  public static final int MAX_OFFSET_FETCH_WAIT_TIME_MS = 5000;
 
-    IngestionInfo(long ingestionTimeMs, long firstStreamIngestionTimeMs,
-        @Nullable StreamPartitionMsgOffset currentOffset, @Nullable StreamPartitionMsgOffset latestOffset) {
+  private static class IngestionInfo {
+    volatile Long _ingestionTimeMs;
+    volatile Long _firstStreamIngestionTimeMs;
+    volatile StreamPartitionMsgOffset _currentOffset;
+    volatile StreamPartitionMsgOffset _latestOffset;
+    final StreamMetadataProvider _streamMetadataProvider;
+
+    IngestionInfo(@Nullable Long ingestionTimeMs, @Nullable Long firstStreamIngestionTimeMs,
+        @Nullable StreamPartitionMsgOffset currentOffset,
+        @Nullable StreamMetadataProvider streamMetadataProvider) {
```

Comment:
`streamMetadataProvider` is also always non-null, and we should be able to remove several unnecessary null checks